torch/csrc/jit/codegen/cuda/expr_simplifier.h

```diff
@@ -0,0 +1,517 @@
+//
+// Common implementations are:
+//
+// trunc div (round to zero):
+// a/b := trunc(a \ b)
+// a%b defined by the fundamental division-with-remainder equation
+//
+// floor div:
+// a/b := floor(a \ b)
+// a%b defined by the fundamental division-with-remainder equation
+//
+// For C89, the result of negative div is not specified. C99 and C++ uses trunc
+// div. Pyton and PyTorch uses floor div. We will only be interested in trunc
+// div here because we use C++.
+//
+// The properties of trunc div are:
+// 1) Good: (-a)/b = -(a/b) = a/(-b)
+// 2) Good: (-a)%b = -(a%b) = a%(-b)
+// 3) Bad: a % b = a' % b is not equivalent to a = a' (mod b)
+//
+// 6. Properties of Div and Mod Under Trunc Div
+// ============================================
+//
+// In this section, I will study trunc div and its properties. I will first
+// redefine trunc div using the same language as in Euclid's division lemma,
+// which will be convenient for proving theorems. I will then prove that this
+// new definition is equivalent to the definition of trunc div as described in
+// section "Implementations of Div and Mod". Then I will study the theorems in
+// section "Some More Theorems" to find out which is true and which needs
+// change. All / and % in this section are using trunc div.
+//
+// Definition 6.0: For any integers a and b (b != 0), there exist unique
+// integers q and r such that
+// 1) if a >= 0, 0 <= r < |b|; if a < 0, -|b| < r <= 0.
+// 2) a = bq + r
+// We can then define a/b := q, a%b := r
+//
+// Theorem 6.0: Definition 6.0 is equivalent to the definition of trunc div in
+// section "Implementations of Div and Mod"
+// Proof: trunc(a\b) is to remove the non-integer portion of a\b. To shift a by
+// r, which obtains a - r, to the nearest multiple of b. To shift towards zero,
+// r and a should have the same sign. Also, |r| must be smaller than |b|
+// otherwise it won't be the nearest multiple of b, and this shift is unique.
+// We therefore proved 1) and 2) above and the uniqueness.
+//
+// Now let's look at theorems in "Some More Theorems" to see how they should be
+// modified when translating to the language of trunc div. Theorems will be
+// numbered consistently, that is, theorem 6.x is the modified version of
+// theorem 4.x.
+//
+// Theorem 6.1: Associativity of trunc div:
+// 1) a*(b/c) != (a*b)/c
+// 2) a/(b*c) = (a/b)/c
+// 3) a/(b/c) != (a/b)*c != (a*c)/b
+// Note that for 2), it is now a "=" instead of a "!=".
+// Proof: for 1) and 3), the same counter example as in Theorem 4.1 applies.
+// For 2), from Definition 6.0, we have
+// a = (a/b)b + a%b = (((a/b)/c)c + (a/b)%c)b + a%b
+//   = ((a/b)/c)*bc + (a%b + ((a/b)%c)*b)
+//   = ((a/b)/c)*bc + (a%b + ((a/|b|)%c)*|b|)
+// if a >= 0, then 0 <= a%b < |b|, 0 <= (a/|b|)%c*|b| <= (|c| - 1)|b|,
+// as a result, we have 0 <= (a%b + ((a/|b|)%c)*|b|) < |bc|,
+// from Definition 6.0, we know that
+// a/(bc) = (a/b)/c and a%(bc) = a%b + ((a/b)%c)*b
```

Comment:
I updated this doc to state this more clearly. It is because we just proved that 
```C++
a = ((a/b)/c)*bc + (a%b + ((a/b)%c)*b)
```
And from Definition 6.0,
```C++
a = (a/(bc))*(bc) + a%(bc)
```
comparing these two equations, due to the uniqueness of the above decomposition, as long as ` (a%b + ((a/b)%c)*b)` falls in the correct range of `a%(bc)` as defined in Definition 6.0, we can state that:
```
(a/(bc)) = (a/b)/c
a%(bc) =  ((a/b)%c)*b)
```

Reply:
Thanks. Now I see that `(a/(bc)) = (a/b)/c`, but still not about the second one. I see this holds: `(a%b + ((a/b)%c)*b) = a%(bc)`, then why `a%b` can be dropped?

Reply:
Oh sorry, it was a copy-paste error. `a%b` can not be dropped. It should be
```
a%(bc) = a%b + ((a/b)%c)*b
```
instead of
```
a%(bc) =  ((a/b)%c)*b)
```
Thanks for pointing it out.

Reply:
Thanks. Makes sense now.